### ISSUE TITLE
GPU node provisioning running into exit code 85 

### ISSUE THREAD

**Issue by vijaygos** [Azure/aks-engine#2406]
We have a cluster-based of AKS-engine version 0.36.4 where the GPU VMSS node pool is not adding any new nodes.
The error is as follows:
VM has reported a failure when processing extension 'vmssCSE'. Error message: "Enable failed: failed to execute command: command terminated with exit status=85 [stdout] [stderr] Connection to k8s.gcr.io 443 port [tcp/https] succeeded! Connection to gcr.io 443 port [tcp/https] succeeded! Connection to docker.io 443 port [tcp/https] succeeded! ".

Other details:
Region: East US

I found a [check-in](https://github.com/Azure/aks-engine/pull/1538) from @CecileRobertMichon that seems to address a similar problem. But I am not a 100% sure if this is a duplicate of the issue the fix was originally intended for.

P.S. I have tried to get the audit and provisioning logs based on the instructions [here](https://github.com/Azure/aks-engine/blob/master/docs/howto/troubleshooting.md). But, the VMSS does not let me reset the password to log into serial console (have another ticket opened for this)

**Comment by CecileRobertMichon** [comment]
@vijaygos could you try scaling with the latest aks-engine version?  ie. update your aks-engine binary version (no need to upgrade the cluster) and try to scale using that.

**Comment by CecileRobertMichon** [comment]
@vijaygos actually just realized it's a vmss cluster so you might need to upgrade the cluster first since the extension is part of the scale set.

You can use --force upgrade if you don't want to change k8s version https://github.com/Azure/aks-engine/blob/master/docs/topics/upgrade.md#forcing-an-upgrade

**Comment by vijaygos** [comment]
@CecileRobertMichon , unfortunately, our upgrade process is a little bit more involved. Today our upgrade process is not designed to upgrade an existing cluster but actually create a new cluster (which involves deploying our bits and transferring live traffic over)

**Comment by jackfrancis** [comment]
O.K., to anyone running into this issue, there is a known workaround.

Specifically, if you are running a GPU-backed VMSS cluster built w/ a version of aks-engine prior to v0.38.0. With a recent version of aks-engine (v0.44.2 is the latest as of this comment):

1. Scale up/out your VMSS node pool by 1, e.g:
```
aks-engine scale --subscription-id <subscription_id>  --api-model _output/<resource_group_name>/apimodel.json --location <region> --resource-group <resource_group_name> --apiserver <resource_group_name>.<region>.cloudapp.azure.com --node-pool <pool_name> --new-node-count 3 --auth-method client_secret --client-id <id> --client-secret <secret>
```
(The above example assumes a 2 node cluster, thus `--new-node-count 3`.)
2. Your VMSS model is now updated, which means any subsequent scale operations via the Azure API (portal/CLI/cluster-autoscaler) will render net new nodes according to the new model provided by the newer version of aks-engine.
3) If you wish to reconcile these changes across the cluster, you may cordon/drain the remaining nodes + add a new node at an appropriate percent-of-nodes interval (e.g., one at a time), eventually yielding a node pool with entirely updated nodes

**Comment by jackfrancis** [comment]
@vijaygos I updated the issue description to reflect the version of aks-engine being v0.36.4, and I confirmed that the fix to this problem was delivered in v0.38.0 via this change:

https://github.com/Azure/aks-engine/pull/1538

The underlying root cause is that older clusters eventually take too long running various background apt operations (e.g., update kernel), which stalls the nvidia installation (which leverages apt) until it times out. The fix in #1538 added an infinite wait for the release of apt locks prior to beginning the nvidia drivers installation, which is what older GPU-backed clusters require.
